File: syntavra_runtime/config_last_good_plan.py
from __future__ import annotations

import argparse
import hashlib
import json
import re
from pathlib import Path
from typing import Any, Mapping, Sequence

from .config_contract import MAX_CONFIG_WIRE_BYTES, decode_config_wire, resolve_config_phases
from .state_snapshot_contract import StateInspectionError, project_id_for_root

SCHEMA_VERSION = 1
CONTRACT_VERSION = 1
PLAN_ID = "syntavra-config-last-good-plan-v1"
TARGET_RELATIVE_PATH = ".syntavra/pre-release/config-last-good.json"
CLAIM = "RUST_CONFIG_LAST_GOOD_PLAN_PARITY_PROVEN_R25_FIXTURES"
PERSISTENT_SCOPES = ("user", "project", "environment")
EPHEMERAL_SCOPES = ("session", "task")
_LOWER_HEX_64 = re.compile(r"^[0-9a-f]{64}$")


class ConfigLastGoodPlanError(ValueError):
    def __init__(self, code: str):
        super().__init__(code)
        self.code = code


def _canonical_payload(snapshot: Mapping[str, Any]) -> bytes:
    expected_keys = {
        "schema_version",
        "values",
        "provenance",
        "config_hash",
        "warnings",
    }
    actual_keys = {str(key) for key in snapshot}
    if actual_keys != expected_keys:
        raise ConfigLastGoodPlanError("CONFIG_LIFECYCLE_SNAPSHOT_KEYS_INVALID")
    return json.dumps(
        dict(snapshot),
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
    ).encode("utf-8")


def _contains_ephemeral_scope(
    phases: Sequence[Mapping[str, Mapping[str, Any]]],
) -> bool:
    return any(bool(phase.get(scope)) for phase in phases for scope in EPHEMERAL_SCOPES)


def config_last_good_plan(
    *,
    project_root: str | Path,
    expected_project_id: str,
    config_wire: bytes,
) -> dict[str, Any]:
    if _LOWER_HEX_64.fullmatch(expected_project_id) is None:
        raise ConfigLastGoodPlanError("CONFIG_LIFECYCLE_EXPECTED_PROJECT_INVALID")

    raw_wire = bytes(config_wire)
    if len(raw_wire) > MAX_CONFIG_WIRE_BYTES:
        raise ConfigLastGoodPlanError("CONFIG_LIFECYCLE_WIRE_TOO_LARGE")

    try:
        actual_project_id = project_id_for_root(project_root)
    except StateInspectionError as exc:
        raise ConfigLastGoodPlanError(
            f"CONFIG_LIFECYCLE_{exc.code.removeprefix('STATE_')}"
        ) from exc
    if actual_project_id != expected_project_id:
        raise ConfigLastGoodPlanError("CONFIG_LIFECYCLE_PROJECT_MISMATCH")

    try:
        phases = decode_config_wire(raw_wire)
    except Exception as exc:
        raise ConfigLastGoodPlanError("CONFIG_LIFECYCLE_WIRE_INVALID") from exc
    if _contains_ephemeral_scope(phases):
        raise ConfigLastGoodPlanError("CONFIG_LIFECYCLE_EPHEMERAL_SCOPE_FORBIDDEN")

    try:
        snapshot = resolve_config_phases(phases)
    except Exception as exc:
        raise ConfigLastGoodPlanError("CONFIG_LIFECYCLE_CONFIG_INVALID") from exc

    payload = _canonical_payload(snapshot)
    warnings = [str(item) for item in snapshot.get("warnings", [])]
    fallback_used = bool(warnings)
    decision = "retain-existing" if fallback_used else "write"

    return {
        "apply_authority": "blocked",
        "candidate": {
            "config_hash": str(snapshot["config_hash"]),
            "payload_bytes": len(payload),
            "payload_sha256": hashlib.sha256(payload).hexdigest(),
            "schema_version": int(snapshot["schema_version"]),
            "warnings": warnings,
        },
        "claim": CLAIM,
        "contract_version": CONTRACT_VERSION,
        "decision": decision,
        "fallback_used": fallback_used,
        "full_product_parity": "FULL_PARITY_NOT_PROVEN",
        "input": {
            "config_wire_bytes": len(raw_wire),
            "config_wire_sha256": hashlib.sha256(raw_wire).hexdigest(),
            "ephemeral_scopes_forbidden": list(EPHEMERAL_SCOPES),
            "format": "R6CFG1",
            "persistent_scopes": list(PERSISTENT_SCOPES),
        },
        "mutation": {
            "database_opened": False,
            "filesystem": False,
        },
        "ok": True,
        "plan_id": PLAN_ID,
        "project_binding": {
            "actual": actual_project_id,
            "expected": expected_project_id,
            "matched": True,
        },
        "project_id": actual_project_id,
        "schema_version": SCHEMA_VERSION,
        "target": {
            "atomic_replace_required": True,
            "file_mode": "0600",
            "relative_path": TARGET_RELATIVE_PATH,
        },
    }


def canonical_plan_json(plan: Mapping[str, Any]) -> str:
    return json.dumps(
        dict(plan),
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
    )


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Produce the no-write R25 config last-good lifecycle plan."
    )
    parser.add_argument("expected_project_id")
    parser.add_argument("project_root", type=Path)
    parser.add_argument("config_wire_hex")
    args = parser.parse_args(argv)
    try:
        wire = bytes.fromhex(args.config_wire_hex)
        result = config_last_good_plan(
            project_root=args.project_root,
            expected_project_id=args.expected_project_id,
            config_wire=wire,
        )
    except (ValueError, ConfigLastGoodPlanError) as exc:
        code = exc.code if isinstance(exc, ConfigLastGoodPlanError) else "CONFIG_LIFECYCLE_WIRE_HEX_INVALID"
        print(code)
        return 2
    print(canonical_plan_json(result))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/syntavra-cli/src/config_last_good_plan.rs
#![forbid(unsafe_code)]

use serde_json::{json, Value};
use syntavra_core::sha256_hex;

use crate::config_contract::{resolve_config_wire, snapshot_json};
use crate::state_snapshot_contract::project_id_for_root;

const SCHEMA_VERSION: u32 = 1;
const CONTRACT_VERSION: u32 = 1;
const PLAN_ID: &str = "syntavra-config-last-good-plan-v1";
const TARGET_RELATIVE_PATH: &str = ".syntavra/pre-release/config-last-good.json";
const CLAIM: &str = "RUST_CONFIG_LAST_GOOD_PLAN_PARITY_PROVEN_R25_FIXTURES";
const MAX_CONFIG_WIRE_BYTES: usize = 256 * 1024;

fn valid_lower_hash(value: &str) -> bool {
    value.len() == 64
        && value
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
}

fn map_project_error(error: &str) -> String {
    let suffix = error.strip_prefix("STATE_").unwrap_or(error);
    format!("CONFIG_LIFECYCLE_{suffix}")
}

fn contains_ephemeral_scope(input: &[u8]) -> Result<bool, String> {
    let text =
        std::str::from_utf8(input).map_err(|_| "CONFIG_LIFECYCLE_WIRE_INVALID".to_owned())?;
    if !text.ends_with('\n') || text.lines().next() != Some("R6CFG1") {
        return Err("CONFIG_LIFECYCLE_WIRE_INVALID".to_owned());
    }
    for line in text.lines().skip(1) {
        let fields = line.split('\t').collect::<Vec<_>>();
        if fields.first() == Some(&"a") && matches!(fields.get(1), Some(&"session") | Some(&"task"))
        {
            return Ok(true);
        }
    }
    Ok(false)
}

fn canonical_snapshot_payload(
    snapshot: &crate::config_contract::ConfigSnapshot,
) -> Result<String, String> {
    let rendered = snapshot_json(snapshot)?;
    let value: Value = serde_json::from_str(&rendered)
        .map_err(|_| "CONFIG_LIFECYCLE_SNAPSHOT_JSON_INVALID".to_owned())?;
    let Some(object) = value.as_object() else {
        return Err("CONFIG_LIFECYCLE_SNAPSHOT_KEYS_INVALID".to_owned());
    };
    let expected = [
        "config_hash",
        "provenance",
        "schema_version",
        "values",
        "warnings",
    ];
    if object.len() != expected.len() || expected.iter().any(|key| !object.contains_key(*key)) {
        return Err("CONFIG_LIFECYCLE_SNAPSHOT_KEYS_INVALID".to_owned());
    }
    serde_json::to_string(&value).map_err(|_| "CONFIG_LIFECYCLE_SNAPSHOT_JSON_INVALID".to_owned())
}

pub fn config_last_good_plan_json(
    project_root: &str,
    expected_project_id: &str,
    config_wire: &[u8],
) -> Result<String, String> {
    if !valid_lower_hash(expected_project_id) {
        return Err("CONFIG_LIFECYCLE_EXPECTED_PROJECT_INVALID".to_owned());
    }
    if config_wire.len() > MAX_CONFIG_WIRE_BYTES {
        return Err("CONFIG_LIFECYCLE_WIRE_TOO_LARGE".to_owned());
    }

    let actual_project_id =
        project_id_for_root(project_root).map_err(|error| map_project_error(&error))?;
    if actual_project_id != expected_project_id {
        return Err("CONFIG_LIFECYCLE_PROJECT_MISMATCH".to_owned());
    }
    if contains_ephemeral_scope(config_wire)? {
        return Err("CONFIG_LIFECYCLE_EPHEMERAL_SCOPE_FORBIDDEN".to_owned());
    }

    let snapshot = resolve_config_wire(config_wire)
        .map_err(|_| "CONFIG_LIFECYCLE_CONFIG_INVALID".to_owned())?;
    let payload = canonical_snapshot_payload(&snapshot)?;
    let warnings = snapshot.warnings.clone();
    let fallback_used = !warnings.is_empty();
    let decision = if fallback_used {
        "retain-existing"
    } else {
        "write"
    };

    let plan = json!({
        "apply_authority": "blocked",
        "candidate": {
            "config_hash": snapshot.config_hash,
            "payload_bytes": payload.len(),
            "payload_sha256": sha256_hex(payload.as_bytes()),
            "schema_version": snapshot.schema_version,
            "warnings": warnings,
        },
        "claim": CLAIM,
        "contract_version": CONTRACT_VERSION,
        "decision": decision,
        "fallback_used": fallback_used,
        "full_product_parity": "FULL_PARITY_NOT_PROVEN",
        "input": {
            "config_wire_bytes": config_wire.len(),
            "config_wire_sha256": sha256_hex(config_wire),
            "ephemeral_scopes_forbidden": ["session", "task"],
            "format": "R6CFG1",
            "persistent_scopes": ["user", "project", "environment"],
        },
        "mutation": {
            "database_opened": false,
            "filesystem": false,
        },
        "ok": true,
        "plan_id": PLAN_ID,
        "project_binding": {
            "actual": actual_project_id,
            "expected": expected_project_id,
            "matched": true,
        },
        "project_id": expected_project_id,
        "schema_version": SCHEMA_VERSION,
        "target": {
            "atomic_replace_required": true,
            "file_mode": "0600",
            "relative_path": TARGET_RELATIVE_PATH,
        },
    });

    serde_json::to_string(&plan).map_err(|_| "CONFIG_LIFECYCLE_PLAN_JSON_INVALID".to_owned())
}

#[cfg(test)]
mod tests {
    use super::{contains_ephemeral_scope, valid_lower_hash};

    #[test]
    fn validates_lowercase_project_ids() {
        assert!(valid_lower_hash(&"a".repeat(64)));
        assert!(!valid_lower_hash(&"A".repeat(64)));
        assert!(!valid_lower_hash("abc"));
    }

    #[test]
    fn rejects_ephemeral_config_scopes() {
        let wire = b"R6CFG1\nphase\t0\na\tsession\t73\t70\ts\t76\n";
        assert_eq!(contains_ephemeral_scope(wire), Ok(true));
    }
}
